storm-core/src/jvm/org/apache/storm/daemon/supervisor/Slot.java

```diff
@@ -0,0 +1,766 @@
+     * PRECONDITION: container != null && currentAssignment != null
+     * @param dynamicState current state
+     * @param staticState static data
+     * @return the next state
+     * @throws Exception on any error
+     */
+    static DynamicState handleWaitingForWorkerStart(DynamicState dynamicState, StaticState staticState) throws Exception {
+        assert(dynamicState.container != null);
+        assert(dynamicState.currentAssignment != null);
+        
+        LSWorkerHeartbeat hb = dynamicState.container.readHeartbeat();
+        if (hb != null) {
+            long hbAgeMs = (Time.currentTimeSecs() - hb.get_time_secs()) * 1000;
+            if (hbAgeMs <= staticState.hbTimeoutMs) {
+                return dynamicState.withState(MachineState.RUNNING);
+            }
+        }
+        
+        if (!equivalent(dynamicState.newAssignment, dynamicState.currentAssignment)) {
+            //We were rescheduled while waiting for the worker to come up
+            return Slot.killContainerForChangedAssignment(dynamicState, staticState);
+        }
+        
+        long timeDiffms = (Time.currentTimeMillis() - dynamicState.startTime);
+        if (timeDiffms > staticState.firstHbTimeoutMs) {
+            LOG.warn("SLOT {}: Container {} failed to launch in {} ms.", staticState.port, dynamicState.container, staticState.firstHbTimeoutMs);
+            dynamicState.container.kill();
+            Time.sleep(staticState.killSleepMs);
+            return dynamicState.withState(MachineState.KILL_AND_RELAUNCH);
+        }
+        Time.sleep(1000);
+        return dynamicState;
+    }
+
+    /**
+     * State Transitions for RUNNING state.
+     * PRECONDITION: container != null && currentAssignment != null
+     * @param dynamicState current state
+     * @param staticState static data
+     * @return the next state
+     * @throws Exception on any error
+     */
+    static DynamicState handleRunning(DynamicState dynamicState, StaticState staticState) throws Exception {
+        assert(dynamicState.container != null);
+        assert(dynamicState.currentAssignment != null);
+        
+        if (!equivalent(dynamicState.newAssignment, dynamicState.currentAssignment)) {
+            LOG.warn("SLOT {}: Assignment Changed from {} to {}", staticState.port, dynamicState.currentAssignment, dynamicState.newAssignment);
+            //Scheduling changed while running...
+            return killContainerForChangedAssignment(dynamicState, staticState);
+        }
+        if (dynamicState.container.didMainProcessExit()) {
+            LOG.warn("SLOT {}: main process has exited", staticState.port);
+            return killAndRelaunchContainer(dynamicState, staticState);
+        }
+        
+        LSWorkerHeartbeat hb = dynamicState.container.readHeartbeat();
+        if (hb == null) {
+            LOG.warn("SLOT {}: HB returned as null", staticState.port);
+            //This should never happen, but to be safe
```

Comment:
As we discussed, this could happen in the case when we successfully recover a container while the worker it contains never initialized far enough to write its first heartbeat. This is still an OK state to be in, so the comment should change.